src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala

```diff
@@ -446,18 +455,43 @@ class GroupReadsByUmi
     * sub-grouping into UMI groups by original molecule.
     */
   def assignUmiGroups(pairs: Seq[ReadPair]): Unit = {
-    val rawUmis = pairs.map(_._1).map(getRawUmi)
-    val rawToId = this.assigner.assign(rawUmis)
-
-    pairs.foreach(pair => Seq(pair._1, pair._2).foreach(rec => {
-      val raw = getRawUmi(rec)
-      val id  = rawToId(raw)
-      rec(this.assignTag) = id
-    }))
+    val umis = pairs.map { case (r1, r2) => umiForRead(r1, r2) }
+    val rawToId = this.assigner.assign(umis)
+
+    pairs.foreach { case (r1, r2) =>
+      val umi = umiForRead(r1, r2)
+      val id  = rawToId(umi)
+      r1(this.assignTag) = id
+      r2(this.assignTag) = id
+    }
   }
 
-  private def getRawUmi(rec: SamRecord): String = rec.get[String](this.rawTag) match {
-    case None | Some("") => fail(s"Record '$rec' was missing the raw UMI tag '${this.rawTag}'")
-    case Some(umi) => umi.toUpperCase
+  /**
+    * Retrieves the UMI to use for a read pair.  In the case of single-umi strategies this is just
+    * the upper-case UMI sequence.
+    *
+    * For Paired the UMI is extended to encode which "side" of the template the read came from - the earlier
+    * or later read on the genome.  This is necessary to ensure that, when the two paired UMIs are the same
+    * or highly similar, that the A vs. B groups are constructed correctly.
+    */
+  private def umiForRead(r1: SamRecord, r2: SamRecord): String = r1.get[String](this.rawTag) match {
+    case None | Some("") => fail(s"Record '$r1' was missing the raw UMI tag '${this.rawTag}'")
+    case Some(chs) =>
+      val umi = chs.toUpperCase
+
+      this.assigner match {
+        case paired: PairedUmiAssigner =>
+          require(r1.refIndex == r2.refIndex, s"Mates on different references not supported: ${r1.name}")
+          val pos1 = if (r1.positiveStrand) r1.unclippedStart else r1.unclippedEnd
+          val pos2 = if (r2.positiveStrand) r2.unclippedStart else r2.unclippedEnd
```

Comment:
See my comment about calling this method twice, since I think getting the unclipped positions isn't the fastest (cigar extraction or alignment blocks?)